File: Haskell/DataStructures/Graph/EulerianCycle.hs
-------------------------------------------------------------------------------
-- Student's name:
-- Student's group:
--
-- Data Structures. Grado en Informática. UMA.
-------------------------------------------------------------------------------

module DataStructures.Graph.EulerianCycle(isEulerian,extractCycle,f, eulerianCycle) where

import DataStructures.Graph.Graph
import Data.List


data Vertex = A | B | C | D | E | F | G | H | I | J deriving (Show,Eq,Enum,Ord)

g3 :: Graph Vertex -- eulerian
g3 = mkGraphEdges [A .. C]
                  [(A,B), (B,C), (C,A)]


--H.1)
isEulerian :: Eq a => Graph a -> Bool
isEulerian g 
	| isEmpty g = False
	| length(vertices g) == 1 = True
	| otherwise = (filter (\x->length x `mod` 2 == 1) (map (successors g) (vertices g))) == []
-- H.2)
remove :: (Eq a) => Graph a -> (a,a) -> Graph a
remove g (v,u) = foldl deleteVertex g' a' 
	where 	
		g' = deleteEdge g (v,u)
		a' = [v|v <- vertices g' , degree g' v ==0]


-- H.3)
extractCycle :: (Eq a) => Graph a -> a -> (Graph a, Path a)
extractCycle g v0 = aux g v0 [v0]
	where aux g v (x:vs)
		|u==x = (remove g (v,u) , x:vs ++ [u])
		|otherwise = aux (remove g  (v,u)) u (x:vs++[u])
			where u = head (successors g v)

-- H.4)
connectCycles :: (Eq a) => Path a -> Path a -> Path a
connectCycles xs (y:ys) | length xs == 0 = (y:ys)
						| (head xs) == y = (y:ys) ++ (tail xs)
						| otherwise = (head xs) : (connectCycles (tail xs) (y:ys))

-- H.5)
vertexInCommon :: Eq a => Graph a -> Path a -> a
vertexInCommon g cycle = head[v | v <- vertices g, elem v cycle]
--
f g v = vertexInCommon g' cycle
	where (g',cycle) = extractCycle g v
-- H.6) 
eulerianCycle :: Eq a => Graph a -> Path a
eulerianCycle g 
	| not(isEulerian g)=error "Grafo no euleriano"
	| otherwise = aux g (head (vertices g)) []
		where 
			aux g v es
				|isEmpty g' = es'
				|otherwise = aux g' v' es'
					where 
						(g',cycle) = extractCycle g v
						v'=vertexInCommon g' cycle
						es'=connectCycles es cycle

File: Java/eulerianCycle/src/dataStructures/graph/EulerianCycle.java
/**
 * Student's name:
 * Student's group:
 *
 * Data Structures. Grado en Informática. UMA.
 */

package dataStructures.graph;

import java.util.Iterator;

import dataStructures.list.*;
import dataStructures.set.Set;

public class EulerianCycle<V> {
    private List<V> eCycle;

    @SuppressWarnings("unchecked")
    public EulerianCycle(Graph<V> g) {
        Graph<V> graph = (Graph<V>) g.clone();
        eCycle = eulerianCycle(graph);
    }

    public boolean isEulerian() {
        return eCycle != null;
    }

    public List<V> eulerianCycle() {
        return eCycle;
    }

    // J.1
    private static <V> boolean isEulerian(Graph<V> g) {
        if(g.numVertices()<2){
        	return true;
        }else{
        	boolean par=true;
        	Iterator <V> it = g.vertices().iterator();
        	while(par && it.hasNext()){
        		if(g.degree(it.next())%2 == 1) par=false;
        	}
        	return par;
        }
    }

    // J.2
    private static <V> void remove(Graph<V> g, V v, V u) {
        g.deleteEdge(v, u);
        for(V ver:g.vertices()){
        	if(g.degree(ver)==0)g.deleteVertex(ver);
        }
    }

    // J.3
    private static <V> List<V> extractCycle(Graph<V> g, V v0) {
           List <V> cycle=new ArrayList<>();
           V v=v0,vt;
           while(!g.successors(v).isEmpty()){
        	   cycle.append(v);
        	   vt=v;
        	   v=g.successors(v).iterator().next();
        	   if(v!=null)remove(g, vt, v);
           }
           cycle.append(v);
		
        return cycle;
    }

    // J.4
    private static <V> void connectCycles(List<V> xs, List<V> ys) {
    	if(xs.isEmpty()){
			for(V v:ys){
				xs.append(v);
			}
		}else{
			V f=ys.iterator().next();
			Iterator <V> it= xs.iterator();
			int i=0;
			while(it.hasNext()&&!it.next().equals(f)){
				i++;
			}
			xs.remove(i);
			it=ys.iterator();
			while(it.hasNext()){
				xs.insert(i,it.next());
				i++;
			}
		}
    }

    // J.5
    private static <V> V vertexInCommon(Graph<V> g, List<V> cycle) {
    		Set<V> set=g.vertices();
			Iterator<V> it = cycle.iterator();
			V v=null;
			boolean encontrado=false;
			//while(it.hasNext() && set.isElem(v=it.hasNext())){
			while(it.hasNext() && !encontrado )	{
				v=it.next();
				if(set.isElem(v) ){
					encontrado=true;
				}
			}
    		return encontrado?v:null;
    }

    // J.6
    private static <V> List<V> eulerianCycle(Graph<V> g) {
    		if(!isEulerian(g)){
				return null;
			}else{
				List<V> eulerian = new ArrayList<>();
				V v =g.vertices().iterator().next();
				do{
					List<V> cycle=extractCycle(g,v);
					connectCycles(eulerian,cycle);
					if(!g.isEmpty()){
						v=vertexInCommon(g,cycle);
					}
				}while(!g.isEmpty());
				return eulerian;
				
			}
    }
}
